### PR TITLE
UnicodeSet L1 Span & Contains Set

### PR DESCRIPTION
**Pull request by EvanJP**
### Remaining [UnicodeSet L1](https://docs.google.com/document/d/1usXoudAnwCx19pQ18qoNMLTAcHW7IuAE6CuosnmtS5s/edit#heading=h.1o8ojscexytf) functions
- span(str, contained)
- span_back(str, contained)
- contains_set(UnicodeSet)
- Refer to ICU4C [`span`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#a81c5da08ac066dcca2f3574352d3cfd0), [`span_back`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#ac7f0dc56611d37663d6650482fca5b5c), and [`containsAll`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#a69e4af354f7ba37f9c9750f824402493) for API  

#### Notes regarding `span/back(str, contained)` 
- After looking through the span conditions, realized they can be resolved easily with just an aptly named boolean parameter in the function (`contained`), where if `true` will check for characters contained within the set, and otherwise if `false`.

### CODE REVIEW

File: components/uniset/src/uniset.rs [hunk 1]
```diff
@@ -143,6 +143,13 @@ impl UnicodeSet {
         self.contains_query(query as u32).is_some()
     }
 
+    /// Returns the inversion list as a slice
+    ///
+    /// Public only to the crate, not exposed to public
```
Comment (zbraniecki) [inline components/uniset/src/uniset.rs:148]: No need to indicate that, it's visible in source and in docs.

Comment (EvanJP) [reply components/uniset/src/uniset.rs]: agreed and changed

File: components/uniset/src/uniset.rs [hunk 2]
```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
             None => false,
         }
     }
+
+    /// Check if the calling Unicodeset contains all the characters of the given UnicodeSet
```
Comment (sffc) [inline components/uniset/src/uniset.rs:192]: UnicodeSet

File: components/uniset/src/uniset.rs [hunk 3]
```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
...
+    /// assert_eq!(example.span("CABXYZ", true), 3);
+    /// assert_eq!(example.span("XYZC", false), 3);
+    /// assert_eq!(example.span("XYZ", true), 0);
+    /// assert_eq!(example.span("ABC", false), 0);
+    /// ```
+    pub fn span(&self, span_str: &str, contained: bool) -> usize {
+        span_str
+            .chars()
+            .take_while(|&x| self.contains(x) == contained)
+            .count()
```
Comment (sffc) [inline components/uniset/src/uniset.rs:254]: @markusicu Can you take a look at this algorithm for `span` (and `span_back` below)?

File: components/uniset/src/uniset.rs [hunk 4]
```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
...
+    /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
+    /// let a_to_d = UnicodeSet::from_inversion_list(vec![65, 69]).unwrap();
+    /// let f_to_t = UnicodeSet::from_inversion_list(vec![70, 85]).unwrap();
+    /// let r_to_x = UnicodeSet::from_inversion_list(vec![82, 88]).unwrap();
+    /// assert!(example.contains_set(&a_to_d)); // contains all
+    /// assert!(!example.contains_set(&f_to_t)); // contains none
+    /// assert!(!example.contains_set(&r_to_x)); // contains some
+    /// ```
+    pub fn contains_set(&self, set: &UnicodeSet) -> bool {
+        if set.as_inversion_list().len() > self.inv_list.len() {
```
Comment (sffc) [inline components/uniset/src/uniset.rs:208]: Why?

`A–B + C–D` has 4 elements in the inversion list, but it is contained within `A–Z`, which has only 2 elements in the inversion list.

File: components/uniset/src/uniset.rs [hunk 5]
```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
...
+            return false;
+        }
+        let s = set.as_inversion_list().len() as f32;
+        let u = self.inv_list.len() as f32;
+        if s * u.log2() < u {
+            set.inv_list.chunks(2).all(|range| {
+                self.contains_range(
```
Comment (sffc) [inline components/uniset/src/uniset.rs:215]: Please use `self.contains_query` directly, instead of calling through `contains_range`

File: components/uniset/src/uniset.rs [hunk 6]
```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
...
+    /// assert!(!example.contains_set(&f_to_t)); // contains none
+    /// assert!(!example.contains_set(&r_to_x)); // contains some
+    /// ```
+    pub fn contains_set(&self, set: &UnicodeSet) -> bool {
+        if set.as_inversion_list().len() > self.inv_list.len() {
+            return false;
+        }
+        let s = set.as_inversion_list().len() as f32;
+        let u = self.inv_list.len() as f32;
+        if s * u.log2() < u {
```
Comment (sffc) [inline components/uniset/src/uniset.rs:213]: Why are you doing this logical fork?  Is it an optimization?  I'm a little weary of converting these numbers to f32 and then doing a log on them.

File: components/uniset/src/uniset.rs (hunk 3 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs:254]: I abstracted the `span` and `span_back` algorithms from ICU4C (but please provide feedback if something seems off):

**ICU4C span** (span_back is just the condition flipped to go backwards)
```c
UChar32 c;
int32_t start = 0, prev = 0;
do {
    U16_NEXT(str, start, length, c);
    if (spanCondition  != contains(c)) {
        break;
    }
} while((prev = start) < length);
return prev
```

**ICU4X span**
```rust
span_str.chars().take_while(|&x| self.contains(x) == contained).count()
```

File: components/uniset/src/uniset.rs (hunk 6 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs:213]: It's an optimization of a search between `O(S * log(U))` vs `O(U)` where `S` is the argument and `U` is the caller.

If `S` is significantly smaller than `U`, then we can perform our search in `O(S * log(U))` by calling our binary search, and else, we just go along every range in `U` and then check to make sure every range in `S` has been covered. 
(E.g. `S = 2, U = 16 ==> 2 log 16 = 8`)

If we are against the idea of calling `contains_range` and don't want to repeat the code from `contains_range`, we can just do the `O(U)` approach.

File: components/uniset/src/uniset.rs (hunk 2 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs]: agreed and fixed

File: components/uniset/src/uniset.rs (hunk 4 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs]: this was meant to be `.size()` but I accidentally changed it to `.len()`

agreed and fixed

File: components/uniset/src/uniset.rs (hunk 5 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs]: agreed and changed

File: components/uniset/src/uniset.rs (hunk 6 above)
Comment (sffc) [reply components/uniset/src/uniset.rs:213]: Most real-life UnicodeSets contain only a few segments, since related characters are grouped together in the code point space.  The biggest sets are probably the [language exemplar characters](https://unicode-org.github.io/cldr-staging/charts/37/by_type/core_data.alphabetic_information.main.html), since they are less likely to be grouped together.  However, even those sets are probably no more than 10-20 ranges.

The other issue is that although your `O(S * log(U))` algorithm is faster on paper, it could potentially have less memory locality.  You have to bounce around the whole array for each lookup, which could mean a weakened ability for the processor to cache intermediate results.  Walking straight forward through a buffer is more efficient than random access due to memory locality.

I'm not sure how expensive f32::log2 is itself.  You can get the integer part of the logarithm by looking at the exponent in the floating point bits.  I assume computing the fraction part of the logarithm requires a bit of churning.  If you want to keep the fork here, consider calling the integer [leading_zeros](https://doc.rust-lang.org/std/primitive.u32.html#method.leading_zeros) function, which is probably faster and smaller code.  It's not as accurate, but it shouldn't matter which algorithm we use near the boundary.

tl;dr: In my opinion, the linear-time search is simple, understandable, and very fast for all real-life uses.  But, I don't feel strongly and am happy to go with the two-algorithm solution if you think it's better.

**Comment by filmil** [comment]
I demand trial by benchmark.

(i.e. where we don't have benchmarks we either shouldn't care about
performance, or should first have some :)

>
>
>

-- 
--
Fuchsia | Component Framework <http://go/tq-teams-cf> || Intl Team
<http://g/fuchsia-intl-team> | Localization
<http://go/ea-the-kea>

**Comment by sffc** [comment]
If we're talking about benchmarks: fork = bigger code, and no fork = smaller code.

File: components/uniset/src/uniset.rs [hunk 7]
```diff
@@ -264,6 +357,38 @@ mod tests {
         assert!(!check.contains_range(&('A'..'A')));
     }
     #[test]
+    fn test_unicodeset_contains_set_slogu() {
```
[the remark this answers is not in the document]
Comment (EvanJP) [reply components/uniset/src/uniset.rs]: Agreed regarding the reliability by refactoring into 2 helper methods being tested, but for now I just removed the test and optimization.

File: components/uniset/src/uniset.rs (hunk 6 above)
Comment (EvanJP) [reply components/uniset/src/uniset.rs]: Agreed and changed regarding removal unless proven necessary/optimal.

I never considered the expense of calling `f32::log2`, and was more of an optimization for "fun" I guess :)
